Sqlite3 database concurrent use & possible speed improvement

Rogier-5 opened the pull request:
I have created these patches to improve concurrent use of minetest with an external application reading the database. E.g. minetestmapper. One patch can also improve the speed or reduce the risk of corruption.

Major changes:
- Add support for Sqlite3 WAL mode. See https://www.sqlite.org/wal.html.
- Increase abort delay if an external application is accessing (reading) the database.

As there are tradeoffs to enabling WAL mode by default, I put it in a separate patch, which can be merged independently, or not.

The possible speed improvement comes from the fact that WAL mode is safer, thus allowing `sqlite_synchronous` to be set to `normal` (1) instead of `full` (2) without increasing the risk of a corrupt database after an OS crash or power failure. [Quote](https://www.sqlite.org/pragma.html#pragma_synchronous):

> With synchronous=FULL in WAL mode, an additional sync operation of the WAL file happens after each transaction commit. The extra WAL sync following each transaction help ensure that transactions are durable across a power loss, but they do not aid in preserving consistency. If durability is not a concern, then synchronous=NORMAL is normally all one needs in WAL mode.

I.e. in WAL mode with `synchronous = normal`, the risk of minor data loss is a little bit higher, but the loss will still not be significant, and an OS crash results in loss of unsaved blocks anyway. Losing a few more because some OS buffers of the WAL file were not completely written out to disk will not be a significant problem to most server administrators if the benefit is better performance.

sfan5 commented:
:+1: for the code

Not sure whether enabling WAL journaling mode for new databased is a good idea though.

nerzhul commented on src/exceptions.h:
```diff
@@ -95,6 +95,11 @@ class SettingNotFoundException : public BaseException {
 	SettingNotFoundException(const std::string &s): BaseException(s) {}
 };
 
+class SettingInvalidException : public BaseException {
```
InvalidSetting

sfan5 replied:
No, this follows the naming scheme.

sfan5 commented on src/database-sqlite3.cpp:
```diff
@@ -162,14 +177,45 @@ void Database_SQLite3::openDatabase()
+	}
+	if (journal_mode != "DELETE" && journal_mode != "TRUNCATE"
+			&& journal_mode != "PERSIST" && journal_mode != "WAL")
+		throw SettingInvalidException(std::string("Invalid value for configuration parameter 'sqlite_journal_mode': '")
```
Other backends also use `DatabaseExceptions` for this kinda stuff, why not use that?

Rogier-5 replied:
It is not really a database problem, but a configuration problem.

I can adapt the other backends as well ? or do you prefer that I use DatabaseException here ?

sfan5 replied:
i'd prefer just using `DatabaseException` here

Rogier-5 replied:
Changed

Megaf commented:
Needs rebase.

paramat commented:
Do you mean squash?

Megaf commented:
@paramat No, rebase, the diff doesnt seem to match the current master. As discussed on #Minetest-Dev at this time.

paramat commented:
Yes, my mistake, often 'no conflicts' is shown but the patch fails to apply.

Rogier-5 commented:
Rebased to latest minetest tree.

@Megaf: Have you been using WAL mode on your server ? If so, I'd like to hear whether you notice(d) any difference/improvement with WAL enabled ?

Megaf commented:
@Rogier-5 Not yet.
Meanwhile, Wouldn't you like to have another good look at your whole code to spot any possibly hidden mistake.

Rogier-5 commented:
@Megaf: You mean you spotted one ?

Megaf commented:
Thanks @sfan5. It would be awesome having this merged. Concept wise at least.
@Rogier-5 I haven't tied this PR yet. At the moment I'm concentrating in something else. I on Minetest holidays, just spotted some PRs that must be merged and need rebase, yours is one of them.

ShadowNinja commented on src/database-sqlite3.cpp:
```diff
@@ -162,14 +180,53 @@ void Database_SQLite3::openDatabase()
 
+	bool created = false;
 	if (needs_create) {
+		created = true;
```
Instead of adding this new `created` variable you just just use the existing `needs_create` variable.

Rogier-5 replied:
Rather superfluous indeed.

ShadowNinja commented on src/util/string.h:
```diff
@@ -245,6 +245,23 @@ inline std::string lowercase(const std::string &str)
+		s2 += toupper(str[i]);
+
+	return s2;
+}
```
Instead of adding this new function just use the existing `lowercase` function.

Rogier-5 replied:
I  used it for stylistic reasons, as SQL keywords are normally written in uppercase.

I'll change it to lowercase.

ShadowNinja commented on src/database-sqlite3.cpp:
```diff
@@ -295,6 +352,27 @@ Database_SQLite3::~Database_SQLite3()
+		} else {
+			SQLOK(rv, "Failed to checkpoint database")
+		}
+	}
```
This can be flattened out to the following:
```C++
	if (rv == SQLITE_BUSY) {
		infostream << "Failed to checkpoint database:"
			<< " the database is in use by another process"
			<< " (but all data is safe!)"
			<< std::endl;
	} else {
		// No need to check if rv is SQLITE_OK, as that's what SQLOK does.
		SQLOK(rv, "Failed to checkpoint database")
	}
```

ShadowNinja commented on src/database-sqlite3.cpp:
```diff
@@ -295,6 +352,27 @@ Database_SQLite3::~Database_SQLite3()
 	FINALIZE_STATEMENT(m_stmt_delete)
 
+	// Try to checkpoint the entire WAL before exit.
+	// so that the database file is self-contained. Hopefully, this also
```
The English here is broken.

Rogier-5 replied:
I'll reformulate

ShadowNinja commented on src/database-sqlite3.cpp:
```diff
@@ -162,14 +180,53 @@ void Database_SQLite3::openDatabase()
+		} else {
+			SQLOK(rv, "Failed to checkpoint database");
+		}
+	}
```
This can be flattened, see my other comment.

Rogier-5 replied:
OK

ShadowNinja commented on src/database-sqlite3.cpp:
```diff
@@ -162,14 +180,53 @@ void Database_SQLite3::openDatabase()
+	}
+	int rv;
+	int rv = sqlite3_wal_checkpoint_v2(m_database, NULL,
+		SQLITE_CHECKPOINT_RESTART, NULL, NULL)));
```
This doesn't compile.  rv is declared and immediately shadowed, and there are too many closing parenthesis on the function.

Rogier-5 replied:
Sorry about that :-(
I thought I had compiled and started this branch again after cutting the long lines.

ShadowNinja commented on src/database-sqlite3.cpp:
```diff
@@ -162,14 +180,53 @@ void Database_SQLite3::openDatabase()
+			std::string("Invalid value for configuration parameter"
+					"'sqlite_journal_mode': '")
+				+ (change_mode ? "" : "default-") + lowercase(journal_mode)
+				+ "', expected: [default-]{delete|truncate|persist|wal}");
```
Omitting braces on such a large block of code is confusing, please add braces to the `if` block.

Rogier-5 replied:
OK.

ShadowNinja commented on src/database-sqlite3.cpp:
```diff
@@ -136,6 +137,23 @@ void Database_SQLite3::endSave() {
+				<< " pages - is another process using the database ?"
+				<< std::endl;
+		m_wal_backlog = new_wal_backlog;
+	}
```
This causes two checkpoints to occur immediately after each other if the commit involves 1000 or more pages.  You should disable auto-checkpoint so that only one checkpoint occurs per commit.

Rogier-5 replied:
IIUC, any checkpoint, whether it happens at commit or upon explicit request, will checkpoint as many pages as possible. That would mean that if a commit is more than 1000 pages, the subsequent explicit checkpoint would be a no-op. 

So the code effectively already operates as if auto-checkpoint is disabled. I might as well disable it explicitly...

I'll also add a separate checkpointing thread. That should eliminate disk IO delays in the server and emerge threads altogether.

ShadowNinja replied:
It does the same thing, but it's less efficient since the second time it still has to check to make sure that everything is up to date.

ShadowNinja commented on src/database-sqlite3.cpp:
```diff
@@ -136,6 +137,23 @@ void Database_SQLite3::endSave() {
+	if (total_size >= 0 && checkpointed >= 0) {
+		int new_wal_backlog = total_size - checkpointed;
+		// Normally, sqlite would checkpoint every 1000 pages.
+		if (m_wal_backlog / 10000 < new_wal_backlog / 10000)
```
This seems like a rather confusing way to write what is mostly essentially equivalent to `(new_wal_backlog > m_wal_backlog + 10000)`.
Is there a reason to round to 10000 frames?

Rogier-5 replied:
The code prints one message each time the WAL backlog has crossed the next multiple of 10000 frames. Each subsequent crossing will usually happen after a number of checkpoints, each adding (much) less than 10000 to the backlog. So, the difference between the previous and the current backlog will rarely if ever exceed 10000.

The purpose is to avoid flooding the log with messages when the backlog has grown large (e.g. one at every checkpoint), while still regularly printing a message if it keeps growing.

The code obviouly needs a clarifying comment.

ShadowNinja commented on src/database-sqlite3.cpp:
```diff
@@ -136,6 +137,23 @@ void Database_SQLite3::endSave() {
+			SQLITE_CHECKPOINT_PASSIVE, &total_size, &checkpointed),
+		"failed to checkpoint database");
+	if (total_size >= 0 && checkpointed >= 0) {
+		int new_wal_backlog = total_size - checkpointed;
```
I think these variables would be much easier to understand if they were called something like `total_frames` and `checkpointed_frames`.

Rogier-5 replied:
Indeed. I'll rename them.

ShadowNinja commented on src/database-sqlite3.h:
```diff
@@ -56,6 +56,7 @@ class Database_SQLite3 : public Database
 
 	std::string m_savedir;
 
+	int m_wal_backlog;
```
`m_last_wal_backlog` would be a more descriptive name.

Rogier-5 replied:
OK.

Rogier-5 commented:
@ShadowNinja 

Thanks for your feedback.

> You need a way to save the synchronous setting per-database. You should also store the journal mode per-database

That's better indeed. I'll change it.

Rogier-5 commented:
Pushed a revised version.

- Changes requested by ShadowNinja
- Added a separate checkpoint thread (for a 70% reduction in block-save time)
- Added (temporary) code to collect statistics about database behavior (CPU time / clock time).

  Statistics are printed when minetest exits, and every `database_statistics_interval` seconds while minetest is running - `database_statistics_interval` **must** be set in minetest.conf !

Benchmark results:
[Test case: run minetest for 1 hour in 'continuous forward' mode, starting with an empty database]

Summary (from a  clock time point of view; i.e. taking database IO delays into account):

compared to delete mode with synchronous == 2, delete mode with synchronous == 1 and wal mode without checkpointing thread are about 10% faster (but wal mode with synchronous == 1 is as safe against system crashes as delete mode with synchronous == 2). Wal mode _with_ checkpointing thread saves 70% time. Considering just CPU time, wal mode is also faster, saving 26% compared to delete mode.

Journal mode | synch. | WAL checkpoint thread | Average block write time | % | Average Transaction time | %
--- | --- | --- | --- | --- | --- | ---
delete | 2 | - |1508µs | 100% | 59519µs | 100%
delete | 1 | - |1020µs | 68% | 53818µs | 90%
wal | 1 | no | 943µs | 62% | 52612µs | 88%
wal | 1 | yes | 299µs | 20% | 17319µs | 29%

Note: in all cases, the standard deviations are rather large - se below.

Note also that using a separate checkpoint thread results in a 77% reduction in worst case transaction time, going down from 1/3 second to 80 ms !

Raw data:
```
============ [3600] journal: delete; synchronous: 2; disable thread: false ===============
############## Database statistics (END) ##################
BeginSave CPU time:        15.65 ±       5.27  (      7.00 ..     121.00)  [n: 1889]
BeginSave Clock time:      21.45 ±     136.23  (      7.00 ..    5412.00)  [n: 1889]
EndSave CPU time:        1478.84 ±    1028.08  (      8.00 ..    5570.00)  [n: 1889]
EndSave Clock time:     39266.94 ±   20907.54  (      8.00 ..  352636.00)  [n: 1889]
LoadBlock CPU time:        29.15 ±      13.86  (     22.00 ..    1685.00)  [n: 210823]
LoadBlock Clock time:      41.10 ±     275.67  (     22.00 ..   37232.00)  [n: 210823]
SaveBlock CPU time:        36.33 ±      26.29  (     15.00 ..     467.00)  [n: 460274]
SaveBlock Clock time:      52.72 ±     349.90  (     15.00 ..   40054.00)  [n: 460274]
Transaction CPU time:   11136.41 ±   12165.89  (    756.00 ..   57659.00)  [n: 1606]
Transaction Clock time: 59519.50 ±   24853.81  (  23433.00 ..  353474.00)  [n: 1606]
Block save CPU time:       88.02 ±     114.39  (     23.00 ..    1411.00)  [n: 1606]
Block save Clock time:   1508.89 ±    4493.34  (     60.00 ..   52500.00)  [n: 1606]
#####################################################
============ [3600] journal: delete; synchronous: 1; disable thread: false ===============
############## Database statistics (END) ##################
BeginSave CPU time:        15.42 ±       4.05  (      8.00 ..      83.00)  [n: 1896]
BeginSave Clock time:      21.34 ±     131.61  (      8.00 ..    4981.00)  [n: 1896]
EndSave CPU time:        1429.52 ±     979.74  (      8.00 ..    5377.00)  [n: 1896]
EndSave Clock time:     34529.69 ±   17092.01  (      8.00 ..  115439.00)  [n: 1896]
LoadBlock CPU time:        29.03 ±      14.08  (     21.00 ..    1922.00)  [n: 216931]
LoadBlock Clock time:      41.27 ±     588.58  (     22.00 ..  247308.00)  [n: 216931]
SaveBlock CPU time:        35.57 ±      26.16  (     14.00 ..     452.00)  [n: 476870]
SaveBlock Clock time:      51.35 ±     328.27  (     14.00 ..   22225.00)  [n: 476870]
Transaction CPU time:   11137.81 ±   11917.09  (    717.00 ..   55303.00)  [n: 1616]
Transaction Clock time: 53818.70 ±   22314.05  (  20159.00 ..  158132.00)  [n: 1616]
Block save CPU time:       77.72 ±      79.73  (     19.00 ..     985.00)  [n: 1616]
Block save Clock time:   1020.78 ±    2495.44  (     53.00 ..   32529.00)  [n: 1616]
##################################################### 
============ [3600] journal: wal; synchronous: 1; disable thread: true ===============
############## Database statistics (END) ##################
BeginSave CPU time:        15.04 ±       3.35  (      7.00 ..      56.00)  [n: 1896]
BeginSave Clock time:      17.59 ±      47.01  (      7.00 ..    2031.00)  [n: 1896]
EndSave CPU time:        2344.05 ±    2036.25  (     17.00 ..    9862.00)  [n: 1896]
EndSave Clock time:     36673.30 ±   18485.14  (     17.00 ..   90181.00)  [n: 1896]
LoadBlock CPU time:        18.03 ±      10.31  (     13.00 ..    2367.00)  [n: 218591]
LoadBlock Clock time:      25.62 ±     240.05  (     14.00 ..   48207.00)  [n: 218591]
SaveBlock CPU time:        26.84 ±      11.35  (     14.00 ..     299.00)  [n: 480635]
SaveBlock Clock time:      40.06 ±     300.05  (     14.00 ..   24811.00)  [n: 480635]
Transaction CPU time:    9636.26 ±   10814.84  (    582.00 ..   49278.00)  [n: 1625]
Transaction Clock time: 52612.04 ±   22599.44  (  17493.00 ..  155992.00)  [n: 1625]
Block save CPU time:       62.61 ±      59.19  (     22.00 ..     849.00)  [n: 1625]
Block save Clock time:    943.07 ±    2034.18  (     54.00 ..   31584.00)  [n: 1625]
##################################################### 
============ [3600] journal: wal; synchronous: 1; disable thread: false ===============
############## Database statistics (END) ##################
BeginSave CPU time:        14.47 ±       4.12  (      7.00 ..      69.00)  [n: 1897]
BeginSave Clock time:      16.44 ±      22.64  (      7.00 ..     820.00)  [n: 1897]
EndSave CPU time:         985.33 ±     913.08  (      7.00 ..    4550.00)  [n: 1897]
EndSave Clock time:      6546.69 ±    4839.66  (      7.00 ..   30855.00)  [n: 1897]
LoadBlock CPU time:        18.19 ±      10.30  (     13.00 ..    2315.00)  [n: 217819]
LoadBlock Clock time:      26.55 ±     242.68  (     14.00 ..   50619.00)  [n: 217819]
SaveBlock CPU time:        27.89 ±      11.47  (     14.00 ..     313.00)  [n: 479678]
SaveBlock Clock time:      39.46 ±     272.81  (     14.00 ..   20173.00)  [n: 479678]
Transaction CPU time:    8271.88 ±   10199.84  (    223.00 ..   46022.00)  [n: 1639]
Transaction Clock time: 17319.53 ±   14071.30  (    222.00 ..   83200.00)  [n: 1639]
Block save CPU time:       45.44 ±      38.57  (     20.00 ..     419.00)  [n: 1639]
Block save Clock time:    299.82 ±     992.29  (     21.00 ..   16537.00)  [n: 1639]
#####################################################
```

Rogier-5 commented:
Fixed compilation on all platforms. For the moment, the (temporary) statistics code will not work on Windows and OSX. The implementation of WAL mode should not be affected at all.

As the statistics code is temporary anyway, I'll try to implement it for other platforms only if asked to.

Fixer-007 commented:
I want to test it in game, if I apply this PR4420, WAL mode will be enabled by default?

Rogier-5 commented:
@Fixer-007: On new databases: yes (unless `sqlite_journal_mode` is set to something different than `wal` in minetest.conf).

Existing databases are not modified unless requested. You'll have to add `sqlite_journal_mode = wal` to the world's `world.mt`. Add `sqlite_synchronous = -1` as well. (See the documentation in minetest.conf.example, or in the advanced options menu).

The wal setting is persistent, so if you want to change an existing database back to `delete` mode after testing, set `sqlite_journal_mode = delete` in `world.mt`, and start minetest once.

If you set `debug_log_level = info`, you'll see messages in debug.txt about which journaling mode is in effect, about the selected synchronous level, and about the starting of the checkpoint thread.

ShadowNinja commented on src/database-sqlite3.cpp:
```diff
@@ -62,88 +68,484 @@ SQLite format specification:
+}
+
+
+void SQLite3::setJournalMode()
```
This isn't a very good name since it looks like a setter.  I suggest `applyJournalMode`.

ShadowNinja commented on src/database-sqlite3.cpp:
```diff
@@ -62,88 +68,484 @@ SQLite format specification:
+}
+
+
+void SQLite3::setSynchronousLevel()
```
Ditto.

ShadowNinja commented on src/database-sqlite3.cpp:
```diff
@@ -62,88 +68,484 @@ SQLite format specification:
+		m_db.checkpointWALPassive();
+		int elapsed = 0;
+		while (!shutdown && elapsed < checkpoint_interval) {
+			checkpoint_interval = g_settings->getU16("sqlite_wal_checkpoint_interval");
```
This should be moved out of the loop.  Fetching settings is fairly slow, and you shouldn't do it more than once if you don't have to.

ShadowNinja commented on src/database-sqlite3.cpp:
```diff
@@ -62,88 +68,484 @@ SQLite format specification:
+bool SQLite3CheckpointThread::shutdown(int timeout)
+{
+	m_shutdown_requested.post();
+	return m_shutdown_acknowledged.wait(timeout * 1000);
```
Why the multiplication?  Just accept the timeout in milliseconds instead.

ShadowNinja commented on src/database-sqlite3.cpp:
```diff
@@ -62,88 +68,484 @@ SQLite format specification:
+void SQLite3::stopWALCheckpointThread()
+{
+	if (m_walCheckpointThread) {
+		if (!m_walCheckpointThread->shutdown(10)) {
```
Ten seconds seems excessive to me.  I think 3-5 would be better.

Rogier-5 replied:
Will change to 5. I don't want to make it too short: I want to avoid forcibly terminating the thread without it having a chance to close the database. The typical shutdown time would be a split second, and IMO the chance of the maximum ever being hit at all are probably nearly zero.

ShadowNinja commented on src/client.cpp:
```diff
@@ -821,9 +823,12 @@ void Client::initLocalMapSaving(const Address &address,
-	m_localdb = new Database_SQLite3(world_path);
+	Settings conf;
+	conf.set("sqlite_journal_mode", g_settings->get("sqlite_journal_mode"));
+	m_localdb = new Database_SQLite3(world_path, conf);
```
This will change the journal mode on existing local maps.  Preferably, it would default to whatever the map was currently using.

Rogier-5 replied:
I assume users (who are mostly not server operators) will prefer the same setting for all locally-saved maps; a setting that can be configured in a single place. If it is configurable per local-save-database, then I think hardly anybody, if anybody at all, will ever use that feature. Also, it means that every local-save-database will need its own `world.mt` to store the current settings, and if ever a user wants to use a different mode, he'll have to check and adapt every `world.mt` of every server he uses.

ShadowNinja commented:
Er, disregard that approval.  I'm not sure how to remove it without adding an outright disapproval.  Consider me neutral ATM.

Rogier-5 commented:
@shadowninja: changes made.

For the database for local map saving, I really think that saving a database-specific setting in `world.mt` will be an inconvenience more often than it will be of use. Nevertheless, I agree that some people would want to configure it for each local database separately. I therefore changed the implementation to use the value from `world.mt` if that file exists, and else fallback to the value from `minetest.conf`. This allows most people to rely on the default setting in `minetest.conf`, while still allowing people who need it to configure it in `world.mt`. Would that be acceptable to you ?

Of course , that does not change the fact that existing databases will automatically change to wal mode unless the user configures them differently. However:
- wal mode is probably what most users/players want anyway, as it significantly increases database IO performance without increasing the risk of corruption
- else many local databases will remain in `delete` mode, just because that's how they were originally created. As a result, many users won't benefit from the better performance.
- it's not _the_ world database, it's only a partial, probably outdated copy on the client.
- even _if_ this ever caused some adverse side-effects to some existing local databases, the real world database will still be on the server, and the local database can easily be 'recovered' from the server.

I moved the statistics code to a separate branch: [sqlite-wal-statistics](https://github.com/Rogier-5/minetest/tree/sqlite-wal-statistics), so that it isn't part of this PR, but still available to anybody who wishes to use it for testing.

ShadowNinja commented:
Seems good, although I haven't looked at the code very closely, nor have I tested this.

Rogier-5 commented:
Rebased to latest minetest tree

paramat commented:
Needs reviews. Note author is inactive for a year, may need adoption.

Megaf commented:
@paramat Understandable, he did a lot of work and this PR for example had no attention from a single dev for over a year.

@Rogier-5 wake up.

paramat commented:
Yes understandable, it needs reviews not author action. My comment is because the author has not been active on Github for a year so may have moved on.

paramat commented:
Quiet for a year, close in 1 month if no commitmnt to adoption.

paramat commented:
Neglected, closing.

paramat commented:
Needs reviews, adoption and rebase.

paramat commented:
No adoption interest shown, closing.
